src/main/kotlin/org/eln2/mc/Extensions.kt

```diff
@@ -0,0 +1,949 @@
+}
+
+inline fun <reified TEntity : BlockEntity> Level.constructMenu(
+    pos: BlockPos,
+    player: Player,
+    crossinline title: (() -> Component),
+    crossinline factory: ((Int, Inventory, ItemStackHandler) -> AbstractContainerMenu),
+    crossinline accessor: ((TEntity) -> ItemStackHandler),
+): InteractionResult {
+
+    return this.constructMenu<TEntity>(pos, player, title) { id, inventory, _, entity ->
+        factory(id, inventory, accessor(entity))
+    }
+}
+
+fun Level.getDataAccess(pos: BlockPos): DataNode? {
+    return ((this.getBlockEntity(pos) ?: return null)
+        as? DataEntity ?: return null)
+        .dataNode
+}
+
+inline fun <reified T : Cell> Level.getCellOrNull(mb: MultiblockManager, cellPosId: BlockPos): T? {
+    val entity = this.getBlockEntity(mb.txIdWorld(cellPosId)) as? CellBlockEntity
+        ?: return null
+
+    return entity.cell as? T
+}
+
+inline fun <reified T : Cell> Level.getCell(mb: MultiblockManager, cellPosId: BlockPos): T =
+    getCellOrNull(mb, cellPosId) ?: error("Cell was not present")
+
+private const val LIBAGE_SET_EPS = 0.001
+fun org.ageseries.libage.sim.electrical.mna.component.Component.connect(pin: Int, info: ElectricalComponentInfo) {
+    this.connect(pin, info.component, info.index)
+}
+
+fun Circuit.add(holder: ComponentHolder<*>) {
+    this.add(holder.instance)
+}
+
+fun Resistor.setResistanceEpsilon(resistance: Double, epsilon: Double = LIBAGE_SET_EPS): Boolean {
+    if (abs(this.resistance - resistance) < epsilon) {
+        return false
+    }
+
+    this.resistance = resistance
+
+    return true
+}
+
+fun VoltageSource.setPotentialEpsilon(potential: Double, epsilon: Double = LIBAGE_SET_EPS): Boolean {
+    if (abs(this.potential - potential) < epsilon) {
+        return false
+    }
+
+    this.potential = potential
+
+    return true
+}
+
+fun Simulator.add(body: ThermalBody) {
+    this.add(body.thermal)
+}
+
+fun Simulator.remove(body: ThermalBody) {
+    this.remove(body.thermal)
+}
+
+fun Simulator.connect(a: ThermalBody, b: ThermalBody, parameters: ConnectionParameters) {
+    this.connect(a.thermal, b.thermal, parameters)
+}
+
+fun Simulator.connect(a: ThermalMass, environmentInformation: EnvironmentInformation) {
+    val connectionInfo = ConnectionParameters(
+        conductance = environmentInformation.airThermalConductivity
+    )
+
+    this.connect(a, environmentInformation.temperature, connectionInfo)
+}
+
+fun Simulator.connect(a: ThermalBody, environmentInformation: EnvironmentInformation) {
+    this.connect(a.thermal, environmentInformation)
+}
+
+/**
+ * This removes the translation I observed in BlockBench models.
+ * Useful for applying transformations like rotation and scale.
+ * */
+fun ModelData.zeroCenter(): ModelData {
+    return this.translate(Vec3(-0.5, 0.0, -0.5))
+}
+
+fun ModelData.blockCenter(): ModelData {
+    return this.translate(Vec3(0.5, 0.0, 0.5))
+}
+
+private const val NBT_ELECTRICAL_RESISTIVITY = "electricalResistivity"
```

Comment:
All the below NBT stuff I think could warrant it's own file as NBTExtensions.kt or something?